### PR TITLE
Add `device.setLocation` support to Android

### LINKED ISSUE #1935: Device.setLocation does nothing on Android
**Describe the bug**

The Device.setLocation API is not implemented on Android. There once was a PR to implement this, but this PR is now locked (or I'd have commented there) and seems to never have been merged: https://github.com/wix/Detox/pull/951

**To Reproduce**

- [X] I have tested this issue on the latest Detox release and it still reproduces

Steps to reproduce:
1. Call `device.setLocation(x, y)` on android
2. Observe nothing happened

**Expected behavior**

The device location should change

**Screenshots**

Not relevant.

**Device and Verbose Detox Logs**

Not relevant.

**Environment (please complete the following information):**
 - Detox: 15.4.0
 - React Native: 0.61.5
 - Node: v10.19.0
 - Device: Any Android Emulator
 - OS: Android
 - Test-runner (select one): Mocha

### PR DESCRIPTION
**Pull request by emilisb**
## Description

- This pull request fixes #1935

In this pull request, I have finished PR #951 from 2.5 years ago which was never merged. I tested the changes locally and they worked fine, I ran the tests many times and they passed every single time.

In addition to Android Emulator, I added Genymotion support. I used [Genymotion Java API](https://docs.genymotion.com/desktop/latest/08_Java_API.html#installing-genymotion-java-api). My implementation should work on both Genymotion Desktop and Genymotion Cloud.

Question for discussion before merging PR: how should we handle a new Genymotion dependency which uses a custom Genymotion Maven repository?

---

> _For features/enhancements:_
 - [x] I have added/updated the relevant references in the [documentation](https://github.com/wix/Detox/tree/master/docs) files.

> _For API changes:_
 - [x] I have made the necessary changes in the [types index](https://github.com/wix/Detox/blob/master/detox/index.d.ts) file.

### CODE REVIEW

**Comment by d4vidi** [comment]
@emilisb this work is amazing, thanks a lot!
We will definitely take the time to discuss and see how to move fwd with the proprietary artifactory.

File: docs/APIRef.DeviceObjectAPI.md [hunk 1]
```diff
@@ -317,11 +316,14 @@ Takes `"portrait"` or `"landscape"` and rotates the device to the given orientat
...
 Check out Detox's [own test suite.](../detox/test/e2e/06.device-orientation.test.js)
 
-### `device.setLocation(lat, lon)` **iOS Only**
+### `device.setLocation(lat, lon)`
 
-> Note: `setLocation` is dependent on [fbsimctl](https://github.com/facebook/idb/tree/4b7929480c3c0f158f33f78a5b802c1d0e7030d2/fbsimctl) which [is now deprecated](https://github.com/wix/Detox/issues/1371). If `fbsimctl` is not installed, the command will fail, asking for it to be installed.
+> Note: 
+> On iOS `setLocation` is dependent on [fbsimctl](https://github.com/facebook/idb/tree/4b7929480c3c0f158f33f78a5b802c1d0e7030d2/fbsimctl) which [is now deprecated](https://github.com/wix/Detox/issues/1371). If `fbsimctl` is not installed, the command will fail, asking for it to be installed.
+> 
+> On Android `setLocation` will work with both Android Emulator (bundled with Android development tools) and Genymotion. The correct permissions must be set in your app manifest.
```
Comment (d4vidi) [inline docs/APIRef.DeviceObjectAPI.md:324]: @emilisb How come we didn't have to add this in our test apps?

File: docs/APIRef.DeviceObjectAPI.md [hunk 2]
```diff
@@ -38,19 +38,18 @@ device.appLaunchArgs.reset();
...
 - [`device.selectApp(name)`](#deviceselectappname)
-- [`device.launchApp()`](#devicelaunchappparams)
+- [`device.launchApp(params)`](#devicelaunchappparams)
 - [`device.terminateApp()`](#deviceterminateapp)
 - [`device.sendToHome()`](#devicesendtohome)
 - [`device.reloadReactNative()`](#devicereloadreactnative)
 - [`device.installApp()`](#deviceinstallapp)
 - [`device.uninstallApp()`](#deviceuninstallapp)
-- [`device.openURL(url)`](#deviceopenurlurl-sourceappoptional)
+- [`device.openURL({url, sourceApp[optional]})`](#deviceopenurlurl-sourceappoptional)
```
Comment (d4vidi) [inline docs/APIRef.DeviceObjectAPI.md:48]: @emilisb Thanks!!! #boyscout

File: detox/src/devices/drivers/android/exec/ADB.js [hunk 3]
```diff
@@ -115,6 +115,28 @@ class ADB {
...
+    // function `strtod` which is locale-sensitive, meaning that depending
+    // on user environment you'll have to send either comma-separated
+    // numbers or dot-separated ones.
+    //
+    // See: https://android.googlesource.com/platform/external/qemu/+/ae0eaf51751391abea2639a65200e724131dc3d6/android/console.c#2273
+    //
+    // As by default Node.js is distributed without ICU, the locale issue
+    // becomes tricky to solve across different platforms, that's why
+    // it's easier for us just to send 2 commands in a row, ignoring one
+    // which will obviosuly fail.
```
Comment (d4vidi) [inline detox/src/devices/drivers/android/exec/ADB.js:129]: 💪🏻

File: docs/APIRef.DeviceObjectAPI.md (hunk 1 above)
Comment (emilisb) [reply docs/APIRef.DeviceObjectAPI.md:324]: It was already there https://github.com/wix/Detox/blob/8c0e4b530e7cb326117f4d64cb56f0a1ee2392d6/detox/test/android/app/src/main/AndroidManifest.xml#L10

Comment (d4vidi) [reply docs/APIRef.DeviceObjectAPI.md:324]: Go figure... Anyways, thanks again 😄

File: detox/src/devices/drivers/android/genycloud/GenyCloudDriver.js [hunk 4]
```diff
@@ -16,6 +16,7 @@ const InstanceLifecycleService = require('./services/GenyInstanceLifecycleServic
 const InstanceLookupService = require('./services/GenyInstanceLookupService');
 const InstanceNaming = require('./services/GenyInstanceNaming');
 const RecipesService = require('./services/GenyRecipesService');
+const DetoxGenymotionManager = require('../../../../android/espressoapi/DetoxGenymotionManager');
```
Comment (noomorph) [inline detox/src/devices/drivers/android/genycloud/GenyCloudDriver.js:19]: Looks fishy a bit, but okay.

**Review by noomorph — APPROVED** [review]
Looks good, just fix the lint :)

**Comment by d4vidi** [comment]
**Finalizing the issue around Genymotion's library**

Taking into consideration that:
1. The library is very small - 52k of code, total.
2. It is hosted on a proprietary artifactory, as Emilis explained, and there is no way to add that repository to hosting apps' build scripts in a seamless way. In order to allow for the library to be downloaded in all apps, we have to introduce a breaking change.
3. Genymotion SaaS usage out there is basically nonexistent, because we've never officially announced it (yet...)

I've taken the liberty to add the library itself as a direct dep of Detox, such that it would be bundled into the `.aar` (or the final test `apk`, in a Detox-compiling app).
This comes alongside a to-do note to remove that and introduce the breaking change in the future major release.

**Comment by d4vidi** [comment]
Check failed due to CI stability. Merging. We must fix that soon...

File: detox/src/devices/drivers/android/genycloud/GenyCloudDriver.js (hunk 4 above)
Comment (d4vidi) [reply detox/src/devices/drivers/android/genycloud/GenyCloudDriver.js]: I agree, but it's mostly about naming. The approach is good.

**Pull request merged**